File: app/src/main/java/com/example/emotionlink/AudioDemo/Client/WebSocketUploader.java
package com.example.emotionlink.AudioDemo.Client;

import android.util.Log;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;
import org.json.JSONException;
import org.json.JSONObject;

import java.net.URI;
import java.util.Base64;
import java.util.Map;

import com.example.emotionlink.AudioDemo.AudioUrlCallback;
import com.example.emotionlink.AudioDemo.WebSocketStatusListener;

public class WebSocketUploader extends WebSocketClient {

    private final AudioUrlCallback audioUrlCallback;
    public WebSocketStatusListener statusListener; // 添加监听器字段
    private int chunkId = 0;
    public void setStatusListener(WebSocketStatusListener listener) {
        this.statusListener = listener;
    }

    public WebSocketUploader(URI serverUri, AudioUrlCallback callback) {
        super(serverUri);
        this.audioUrlCallback = callback;
    }

    @Override
    public void onOpen(ServerHandshake handshakedata) {
        Log.d("WebSocket","Audio WebSocket opened");
        if (statusListener != null) {
            statusListener.onConnected();
        }
    }

    public void sendInit() {
        try {
            JSONObject init = new JSONObject();
            init.put("msg_type", "init");
            init.put("dtype", "int16");
            init.put("sample_rate", 16000);
            init.put("channels", 1);
            send(init.toString());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    public void sendAudioChunk(byte[] audio, int length) {
        try {
            JSONObject audioObj = new JSONObject();
            audioObj.put("msg_type", "chunk_info");
            audioObj.put("chunk_id", chunkId);
            audioObj.put("status", "start"); // 可选："start" 或 "end"
            audioObj.put("chunk_size", length);
            send(audioObj.toString()); // 发送 JSON 元数据

            // 发送音频二进制 chunk（直接发裸 byte[]）
            byte[] chunk = new byte[length];
            System.arraycopy(audio, 0, chunk, 0, length);
            send(chunk); // 发送音频数据本体

            chunkId++; // 递增 chunk_id
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onMessage(String text) {
        try {
            Log.d("Websocket","Audio进入回调"+text);
            JSONObject json = new JSONObject(text);
            if (json.has("data") && !json.isNull("data")) {
                JSONObject data = json.getJSONObject("data");

                String fromUser = data.getString("from_user");
                String fromLanguage = data.getString("from_language");
                String language = data.getString("language");
                String textContent = data.getString("text");
                String wavFile = data.getString("wav_file");

                // 打印字段或进行后续逻辑处理
                System.out.println("fromUser = " + fromUser);
                System.out.println("fromLanguage = " + fromLanguage);
                System.out.println("language = " + language);
                System.out.println("text = " + textContent);
                System.out.println("wavFile = " + wavFile);
            } else {
                System.out.println("data 字段为空或不存在");
            }
            if ("audio".equals(json.optString("type"))) {
                String audioUrl = json.optString("audio_url");
                if (audioUrl != null && !audioUrl.isEmpty() && audioUrlCallback != null) {
                    audioUrlCallback.onAudioUrlReceived(audioUrl);
                }
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }
    public void sendEnd() {
        try {
            JSONObject endObj = new JSONObject();
            endObj.put("msg_type", "chunk_info");
            endObj.put("chunk_id", chunkId); // 最后一个 ID
            endObj.put("status", "end");
            endObj.put("chunk_size", 0);
            send(endObj.toString());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onError(Exception ex) {
       Log.e("Websocket","Audio WebSocket error: " + ex.getMessage());
        if (statusListener != null) {
            statusListener.onError(ex);
        }
    }

    @Override
    public void onClose(int code, String reason, boolean remote) {
        Log.d("Websocket","WebSocket closed. Code: " + code + ", Reason: " + reason + ", Remote: " + remote);

    }
}

File: app/src/main/java/com/example/emotionlink/Repository/WebSocketUploader.kt
package com.example.emotionlink.Repository

import android.util.Log
import com.example.emotionlink.AudioDemo.WebSocketStatusListener
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import org.json.JSONObject
import java.util.concurrent.TimeUnit
import kotlin.concurrent.thread
import kotlin.io.encoding.ExperimentalEncodingApi

class WebSocketUploader(
    wsUrl: String,
    private val messageCallback: MessageCallback
) {
    companion object {
        private const val TAG = "WebSocketUploader"
    }

    private var isReconnecting = false
    private var webSocket: WebSocket? = null
    private var statusListener: WebSocketStatusListener? = null
    private var chunkId: Int = 0
    private var heartbeatJob: Job? = null
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(0, TimeUnit.SECONDS) // 保持长连接
        .build()
    val request = Request.Builder()
        .url(wsUrl)
        .build()

    fun connect() {

        webSocket = client.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                Log.d(TAG, "WebSocket connected")
                statusListener?.onConnected()
                startHeartbeat()
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                Log.d(TAG, "Received: $text")
                handleWebSocketMessage(text, messageCallback)
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {

                Log.d(TAG, "Received binary message: ${bytes.hex()}")
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                Log.e(TAG, "WebSocket error: ${t.message}")
                statusListener?.onError(Exception(t))
                reconnect()
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "WebSocket closing: $reason")
                webSocket.close(code, reason)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                Log.d(TAG, "WebSocket closed: $reason")
            }
        })
    }

    private fun startHeartbeat() {
        heartbeatJob = CoroutineScope(Dispatchers.IO).launch {
            while (isActive) {
                delay(10_000)
                Log.d(TAG, "发送ping")
                send("ping")
            }
        }
    }

    fun send(message: String) {
        webSocket?.send(message)
    }

    private fun reconnect() {
        if (isReconnecting) return
        isReconnecting = true
        CoroutineScope(Dispatchers.IO).launch {
            delay(5_000)
            Log.d(TAG, "尝试重连")
            webSocket = null
            connect()
            isReconnecting = false
        }
    }

    fun setStatusListener(listener: WebSocketStatusListener) {
        statusListener = listener
    }

    fun sendInit() {
        try {
            val init = JSONObject().apply {
                put("msg_type", "init")
                put("dtype", "int16")
                put("sample_rate", 16000)
                put("channels", 1)
            }
            webSocket?.send(init.toString())
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun sendAudioChunk(audio: ByteArray, length: Int) {
        try {
            val meta = JSONObject().apply {
                put("msg_type", "chunk_info")
                put("chunk_id", chunkId)
                put("status", "start")
                put("chunk_size", length/2)//1280
            }
            webSocket?.send(meta.toString())

            val chunk = audio.copyOfRange(0, length)
            webSocket?.send(ByteString.of(*chunk))
            Log.d(TAG, "正在发送")
            chunkId++
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun sendEnd(duration: String,inCancelZone: Boolean) {
        try {
            val status = if (inCancelZone) "discard" else "end"
            val end = JSONObject().apply {
                put("msg_type", "chunk_info")
                put("chunk_id", "-1")
                put("status", status)
                put("chunk_size", 0)
                put("duration", duration)
            }
            Log.d(TAG, "发送结束")
            webSocket?.send(end.toString())
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun close() {
        heartbeatJob?.cancel()
        webSocket?.close(1000, "Normal closure")
    }

    @OptIn(ExperimentalEncodingApi::class)
    fun handleWebSocketMessage(message: String, callback: MessageCallback) {
        try {
            val json = JSONObject(message)

            val data = json.optJSONObject("data")
            if (data != null) {
                val fromLang = data.optString("from_language")
                val toLang = data.optString("language")
                val text = data.optString("text")
                val wavFile = data.optString("wav_file")
                val duration = data.optString("duration")

                if (!text.isNullOrEmpty() || !wavFile.isNullOrEmpty()) {
                    callback.onTextMessageReceived(fromLang, toLang, text, wavFile, duration)
                }
            }
        } catch (e: Exception) {
            callback.onError(e)
        }
    }
}
